### ISSUE TITLE
Support OrangeC attributes in the C++ standard attribute syntax

### ISSUE THREAD

**Issue by chuggafan** [LADSoft/OrangeC#379]
Basically, in the C++ standard (and in the C2x draft, notably, as seen in the post london 2019 docs http://www.open-std.org/jtc1/sc22/wg14/www/docs/PostLondon2019.htm ( http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2385.pdf for the actual draft) the "standard" attribute way is to do namespace-based attributes, e.g. gcc will have gnu::weak or clang will have clang::noescape, so what we should have is the ability to do the OrangeC specific attributes in a namespace such as occ::zstring for example and that way we can have multiple ways of writing the same thing, but use it as the "standard"

https://en.cppreference.com/w/cpp/language/attributes for reference.

This isn't an immediate issue and I can understand if this if it isn't something easily done due to the internals of the code (also why in the nine hells did WG14 add the ability for \_\_x\_\_ to be the same attribute as x when C++ does not do this? Hopefully a later working draft changes this to be more in like with c++).

**Comment by GitMensch** [comment]
> why in the nine hells did WG14 add the ability for __x__ to be the same attribute as x when C++ does not do this? Hopefully a later working draft changes this to be more in like with c++

As quoted before - [GCC docs, Attribute-Syntax](https://gcc.gnu.org/onlinedocs/gcc/Attribute-Syntax.html#Attribute-Syntax)):

> You may also specify attributes with ‘__’ preceding and following each keyword. This allows you to use them in header files without being concerned about a possible macro of the same name. For example, you may use `__noreturn__` instead of `noreturn`.

Therefore I absolutely agree with WG14 to add this ability.

**Comment by LADSoft** [comment]
fwiw last night I added the __attribname__ syntax for gcc, maybe I should have made that clear when I closed the related issue.   the issue brought up here is another issue entirely...

**Comment by GitMensch** [comment]
> the issue brought up here is another issue entirely...

I'm aware of both the closing reason and the difference of this issue.
My comment was a response to @chuggafan as he questioned the usefulness of `__x__` and the GCC docs give the exact reason for it.

**Comment by chuggafan** [comment]
Well the reason I was saying why they were doing it was because of the fact that C++ does not do this so if someone decides to write in a C header [[\_\_maybe_unused\_\_]] it'll not compile in C++, I suspect that the next meeting for WG14 they should have a proposal to move the attributes syntax to be more in line with C++ to prevent such things from occurring.

**Comment by GitMensch** [comment]
Actually you could try to contact your national standard body (or better: simply the WG14) and ask to send a note to the WG for C++ and ask them to add it there, too.

Actually I'm not sure, but should the following work in standard C++?

```cpp
extern "C" {
#include "header_with_double_underscores.h"
}
```

**Comment by LADSoft** [comment]
the __attribute__((__attribname__)) format would still work in a c++ program the way I implemented it.   Should be fine even WITHOUT the extern "C" {}..

I don't think it would be a big deal to support the C++ scoped attributes, the same function in the source code is used to gather both sets of attributes.   When I added the GNU __attribute__ keyword check I added it to the same function that did the C++ [[]] check.    Then generalized the whole thing into an attributes structured stored along with each symbol.   And added the infrastructure to make sure the symbols are updated properly and the attributes are acted on at the right time.   So it is just a matter of parsing out the scoped name then setting up the data structure appropriately, which can easily be done without reference to symbol tables and such since the scoped names are known in advance.    This is actually well-bounded and should be easy enough for someone other than me to do (hint hint).

**Comment by chuggafan** [comment]
I've now found out that OCC doesn't *ACTUALLY* have support for [[noreturn]] in that syntax, whoops, that's a bad oversight (and for that matter it accepts [[carries_dependency]] and also seems to do nothing with it). The only [[attribute]] in that syntax that's actually accepted is [[deprecated]], so I'll fix that up as well, I currently made some changes specifically to use unordered_maps of the files we're searching for (since these are constant and not actually being mutated we're good).

I'll solve that as well, I am going to be unable to do much in the following weeks however so don't expect anything quickly from me, but I'll work the best I can.

**Comment by LADSoft** [comment]
that is fine, we can leave this open until you have time to work on it :smile:

**Comment by chuggafan** [comment]
Looks like I might have been wrong, I was confused as to where attributes were being applied to symbols because an empty symbol doesn't get attributes applied to it until far after in something unrelated by the function AttributeFinish, but as far as I can tell that only applies the CopyFrom vectorsize and cleanup attributes.

And the [[noreturn]] still doesn't get done because it's mixed in with [[carries_dependency]] and then just marks the attribute as special. Also basisAttributes has NOTHING that allows it to easily mark something as noreturn because basisAttribs only has linkage2 not linkage3 and I'm not even sure that the linkage2 in basisAttribs does anything. 

Other than that mess, I've created 3 new attribute related errors (Attribute namespace is not an attribute, attribute (name) does not exist, and attribute (name) does not exist in namespace (name). I switched all the maps used for attributes to constant unordered maps (we do no insertion and so that makes the most sense to me, also unordered maps are for searching them), I also swapped out the operator[] usage on them (which would insert if the item doesn't exist) to using .find and then checking if it exists so far.  I'm continually unsure if there's an "easy" way to implement this, I believe the probable best way to do it is to essentially take a mix of __declspec code and just directly start applying the linkage qualifiers but as far as I can tell there's literally no simple way to do this, and the literal only time I can find a Symbol or Type's linkage to be set is in the function declare. So.... yhea.... no easy way to navigate the code, should be fun to restructure it entirely so that I can get something reasonable out of it... hopefully....

**Comment by chuggafan** [comment]
Hmm, doing further testing it seems that there's no error issued for the function not returning it has the C++ noreturn attribute and is an int returning function but does if there isn't one, I'm confused because as far as I can tell this attribute is quite literally, *NEVER HANDLED* except to mark it as a special attribute. I'm confused now as to how and why. because as I show about, to the best of my knowledge literally nothing is done with it and it gets applied to absolutely nothing, maybe the stuff in declare actually does something? I don't know but I'm confused since the attribute handling is all over the place.

**Comment by chuggafan** [comment]
@GitMensch I got a reply! So I'm just going to summarize what the email said:

Double underscores were made available specifically so that old macros in code could still be used, and it is allowed for the attributes to be accepted by C++ and be conforming already, also 1/2 of the major compilers will accept the double underscores (along with this link: https://godbolt.org/z/BxOEmt ). And maybe user pressure will make them accept double underscores in the future in C++ for the remaining (clang and ICC).

So basically the reason for it is to prevent users from getting broken code due to macros, and so they added it and C++ implementors can do with that what they will. I vote to just make a change to allow the double-underscore style in C++ as well for simplicity and compatibility in that case and then once I can figure out how we actually implement attributes we can add a C2x mode and just enable attributes there and add features for the C2x mode as we will.
